broadinstitute/single_cell_portal_core
# Integration for differential expression PAPI jobs (SCP-4194)

This update adds basic command-line integration of launching differential expression jobs from the Rails console using the `PapiClient` class.  This will now take the various inputs required and submit a job via the Pipelines API, which will localize files from the study bucket, run the calculations, and the delocalize outputs back to the bucket.

In addition, this adds test coverage for the `PapiClient` class to ensure against future updates breaking our PAPI submission infrastructure.  Also, extra parameters for differential expression jobs are now managed by the `DifferentialExpressionParameters` class, which will handle validating and setting CLI options for PAPI.

MANUAL TESTING
To fully validate that you can run a DE job on actual data, this will require creating a new study using the following example data (taken from [this](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP294/de-test-2) study on staging).  The example data can be downloaded from here:

[https://console.cloud.google.com/storage/browser/broad-singlecellportal-staging-testing-data/SCP294_DE](https://console.cloud.google.com/storage/browser/broad-singlecellportal-staging-testing-data/SCP294_DE?project=broad-singlecellportal-staging&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false)

You will need to ingest at least the metadata, clustering, and raw counts matrix for this example to work.
1. Pull this branch, boot as normal, and start `Delayed::Job` workers
2. Set your `Ingest Pipeline Docker Image` to `gcr.io/broad-singlecellportal-staging/scp-ingest-jlc_delocalize_de:dce674337152`
3. Create a new study using the above example data, and ingest required files (this should take 5-10 minutes total)
   *  You can get expression matrix information from `file_supplemental_info.tsv`, though it is not required for this to succeed
4. Once the data is ingested successfully, open up the "Study Details" page for this study and click the link to open the bucket in a browser
5. In a separate rails console session, run the following snippet to assemble the required parameters for the job, interpolating in the accession for your new study:
```
study = Study.find_by(accession: <new study accession>)
cluster_file = study.cluster_ordinations_files.first
metadata_file = study.metadata_file
user = study.user
action = :differential_expression
matrix_file = study.expression_matrix_files.first
options = {
  annotation_name: 'cell_type__ontology_label',
  annotation_type: 'group', 
  annotation_scope: 'study',
  annotation_file: metadata_file.gs_url,
  cluster_file: cluster_file.gs_url,
  cluster_name: cluster_file.name,
  matrix_file_path: matrix_file.gs_url,
  matrix_file_type: 'dense'
}
de_params = DifferentialExpressionParams.new(options)
``` 
6. Submit the job to PAPI:
```
client = ApplicationController.papi_client
job = client.run_pipeline(study_file: cluster_file, user: user, action: action, params_object: de_params)
```
7. To poll for completion, you can re-run the following snippet to ensure that the job completes:
```
pipeline = client.get_pipeline(name: job.name)
pipeline.done?
pipeline.error
```
8. Once `pipeline.done?` returns `true`, confirm that `pipeline.error` returns `nil`
9. Go back to the storage browser tab and validate that the DE output files appear in the bucket:
![Screen Shot 2022-05-04 at 3 39 26 PM](https://user-images.githubusercontent.com/729968/166812978-686a9cae-fe41-4dae-8f2c-d4bc55661d04.png)

This PR satisfies SCP-4194.
